### DISCUSSION TITLE
CI: how to check which apps have changed

### DISCUSSION THREAD

**Discussion by bobaaaaa** [vercel/turborepo#800] [Help] [answered]
Is there a preferred way to check in CI which apps have changed? In our setup we build, test, lint and build our docker image in CI and push it to AWS ECR. 

Anyway, I only want to build the docker images of the changed apps (or a related package).

On github actions there is the [if conditional](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idif). So it would be perfect if there is something like:

```yml
name: example-workflow
on: [push]
jobs:
  production-deploy:
    if: ${{ yarn turbo changed --scope=web --include-dependencies }}
```

Idk if this would work 😬

**Comment by samchouse** [comment]
There should be a --since flag on the run command

**Comment by graup** [comment]
Since there's already the `--since` flag, it would be easy to add a command like you suggest that just returns the list of changed packages. I imagine this could be useful.

**Comment by bobaaaaa** [comment]
In our CI there is an explicit docker build step for each app. The `turbo run --since` task will not help here. I will not run a turbo task. I would like to know if the app has changed or get a list of changed apps.

Or how would you do that with `turbo run --since`?

**Reply by weyert** [reply]
I think you want the `--plan` command for this which is not available yet. I am waiting for that one too. I hope @jaredpalmer picks it up swiftly :)

See also: https://github.com/vercel/turborepo/issues/271

**Reply by bobaaaaa** [reply]
@weyert ah yes. Thank you for the hint ❤️. That would work, I guess.

**Comment by bobaaaaa** [comment]
See #380 (https://github.com/vercel/turborepo/releases/tag/v1.1.7)

> Implement `--dry-run` and `--dry-run=json`

**Reply by furkimon** [reply]
turbo does not have the "changed" script, what does this example mean? how do you detect it exactly?

**Reply by bobaaaaa** [reply]
@furkimon pls take a look at this post: https://github.com/vercel/turbo/discussions/800#discussioncomment-3566719

**Comment by graup** [comment]
@jaredpalmer do you have an example of how to use this as Ignored Build Step in Vercel?

Something like

    yarn turbo run build --dry --since=origin/main | grep "package-to-deploy"; test $? -eq 1

?

**Reply by weyert** [reply]
Yeah, but these days it would be `--filter=[origin/main]` now that `--filter` has been introduced. Only it isn't working for me at the moment: 
https://github.com/vercel/turborepo/issues/1032

**Reply by graup** [reply]
I see that there's some documentation here https://vercel.com/docs/concepts/git/monorepos but it seems outdated. @jaredpalmer could you guys update this with the new filter syntax and also add an example for Ignore Build Step?

**Reply by jaredpalmer** [reply]
- `turbo-ignore` is a highly experimental cli package for the Ignored Build Step we are incubating. You would set the value to `npx turbo-ignore`. Note: the name of this package may change, so definitely watch the repo and for release notes. 
- You would also need to tell Vercel to deep clone by setting an env var to `VERCEL_DEEP_CLONE=1`

**Reply by mrmckeb** [reply]
Just a quick note that `VERCEL_DEEP_CLONE` is an experimental flag and will be removed at some point in the future. The value should also be `VERCEL_DEEP_CLONE=true` :)

**Reply by jarrodjperez** [reply]
Seems like `VERCEL_DEEP_CLONE` is no longer supported. Our builds started failing a few days ago with the error below. We tried setting it to both `1` and `true`, but with no luck. Once we removed it, the error went away.

```
[11:07:14.357] Error: Cloning failed
[11:07:14.357]     at retries (/var/task/sandbox.js:317:5217)
[11:07:14.357]     at async _Ze (/var/task/sandbox.js:317:4839)
```

**Reply by mrmckeb** [reply]
There may have been some changes made to this behaviour, I've raised this with the team @jarrodjperez. Apologies for any inconvenience here.

**Reply by jarrodjperez** [reply]
@mrmckeb thanks for the reply! No worries, as we knew it was experimental when we started using it. Just wish we got a heads up before the behavior changed. For the time being, we moved off `npx turbo-ignore` but would love to move back to it once this is resolved.

**Comment by ChristianIvicevic** [comment]
To add to this discussion I highly recommend looking into [`nrwl/nx-set-shas`](https://github.com/nrwl/nx-set-shas). During the workflow they deduce via the API what the last successful run was and obtain the ref for it. This allows me to do a dry run with `--filter=scope...[$NX_HEAD...$NX_BASE]` and ensure that all affected packages are correctly built across multiple commits and changes.

**Comment by bobaaaaa** [comment, marked as answer]
Here is an example how we do this on github actions:

```yml
name: Pipeline

# ... more global settings (not relevant here)

jobs:
  checks:
    name: Build
    # ... more job settings (not relevent here)
    env:
      TURBO_RUN_FILTER: ${{ github.event_name == 'pull_request' && '...[origin/main]' || '...[HEAD^]' }}
    outputs:
      # Defining a job output for used by the next jobs:
      # https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idoutputs
      changed-app: ${{ steps.changed-app.outputs.result }}
      changed-docs: ${{ steps.changed-docs.outputs.result }}

    steps:
    # ... all basic steps you need (checkout etc.)

      - name: Changeset
        id: changeset
        # 1. We need the 'output' of a turbo dry-run to get a json with all affected packages of these run.
        # 2. The multi line json string is transformed to a single line string.
        # 3. The single line json string is set to an 'output' (or result) of this step.
        run: |
          content=`yarn build --filter=$TURBO_RUN_FILTER --dry-run=json`
          # the following lines are required for multi line json
          content="${content//'%'/'%25'}"
          content="${content//$'\n'/'%0A'}"
          content="${content//$'\r'/'%0D'}"
          echo "::set-output name=result::$content"

      - name: Changed app?
        id: changed-app
        # 1. Using the 'output' of the 'changeset' step inside the `if` to check if the specific app has changed
        # 2. When `true`, we store it in the 'output' of the current step.
        if: ${{ contains(fromJSON(steps.changeset.outputs.result).packages, 'app') }}
        run: |
          echo "::set-output name=result::true"

      - name: Changed docs?
        id: changed-docs
        # 1. Using the 'output' of the 'changeset' step inside the `if` to check if the specific app has changed
        # 2. When `true`, we store it in the 'output' of the current step.
        if: ${{ contains(fromJSON(steps.changeset.outputs.result).packages, 'docs') }}
        run: |
          echo "::set-output name=result::true"

  publish_app:
    name: Publish app
    needs: [checks]
    # ... more job settings (not relevent here)
    if: ${{ needs.checks.outputs.changed-app }}
    # Now this job only runs when the 'output' of the previous step is `true` for `changed-app`

  publish_docs:
    name: Publish docs
    needs: [checks]
    # ... more job settings (not relevent here)
    if: ${{ needs.checks.outputs.changed-docs }}
    # Now this job only runs when the 'output' of the previous step is `true` for `changed-docs`
```

I tried a lot to achieve this (env vars etc.). If anyone has a simpler and not so error-prone solution, I would love to see.

@jaredpalmer maybe the team can check my solution? I think a custom turborepo github action would be a nice tool here.

**Reply by elliottsj** [reply]
The escaping for multi-line JSON is no longer necessary; can do this instead:

```yaml
        run: |
          echo 'RESULT<<EOF' >> $GITHUB_OUTPUT
          yarn build --dry-run=json >> $GITHUB_OUTPUT
          echo 'EOF' >> $GITHUB_OUTPUT
```

`${{ contains(fromJSON(steps.changeset.outputs.RESULT).packages, 'app') }}`

Context:
- https://stackoverflow.com/a/61919791
- https://github.com/actions/toolkit/issues/403
- https://github.com/actions/toolkit/blob/b36e70495fbee083eb20f600eafa9091d832577d/docs/commands.md#set-an-environment-variable

**Reply by bobaaaaa** [reply]
Ah cool. Thx 👍

**Reply by jplwood** [reply]
Does this actually determine which packages have changed? Looking at the output of a `--dry-run=json` seems to simply indicate all packages in the build, regardless if they've changed or not. I assume this needs a smarter mechanism using `--filter` to determine what changed between the base branch and the current HEAD. Similar to what @alexbchr is doing internal to his action https://github.com/vercel/turbo/discussions/800#discussioncomment-4238541

**Reply by bobaaaaa** [reply]
You're right with the `--filter` flag. Everything else works as expected. In our prod pipeline.

```
env:
  TURBO_RUN_FILTER: ${{ github.event_name == 'pull_request' && '...[origin/main]' || '...[HEAD^]' }}

yarn build --filter=$TURBO_RUN_FILTER --dry-run=json
```

I updated my original post

**Reply by alexbchr** [reply]
Yeah @jplwood without the `--filter` arg the `--dry-run=json` is simply gonna print all packages that are targeted by your `build` pipeline.

**Reply by jplwood** [reply]
For posterity, I just wanted to follow up on this with more details, building on the setup that @bobaaaaa is showing above! 

First, I wanted to share that I landed on a slightly different, more dynamic filter for determining changes: 

If the turbo command is `npx -y turbo build --dry-run=json --filter=...[$TURBO_REF_FILTER]`, `$TURBO_REF_FILTER` is:

```yaml
TURBO_REF_FILTER: ${{ github.event_name == 'pull_request' && github.event.pull_request.base.sha || github.event.before }}
```

Where 

- If it’s a PR (i.e., `github.event_name == 'pull_request'`), then `$TURBO_REF_FILTER` will be `github.event.pull_request.base.sha` or the HEAD commit of the branch you’re merging into
  - See [webhook-events-and-payloads#pull_request](https://docs.github.com/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#pull_request) object payload
  - I was worried about using `origin/main` for PRs in case I'm making a PR against a different branch, I still want this command to result in a correct output.
- If it's being run in the context of a merge (i.e., on: push), `$TURBO_REF_FILTER` will be `github.event.before` or the HEAD commit SHA on your target branch before the merge.
  - See [webhook-events-and-payloads#push](https://docs.github.com/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#push) object payload
  - If I used `HEAD^` here instead, unless I'm mistaken, that assumes rigid use of squash merges into your main branch, if someone accidentally does a regular merge and they have more than one commit, I don't think `HEAD^` will point to the right thing.

----

Second, I put all of this into a dedicated GitHub actions workflow that I can reuse across CI and CD (Note, I admit there's likely a way to make this more terse, but my GHA skillz are pretty basic. Also, the workflow's `outputs` are specific to _my_ app package names):
<details>
  <summary><code>.github/workflows/changed-packages.yml</code></summary>

Note, the name of my apps are `api` and `ui` -- the `--dry-run` JSON checks and ouputs you use will need to refer to the names of _your_ packages.

```yml
# Turbo dry run to determine which apps have changed between the latest commit and target or previous commit
name: 'Determine changed packages'

on:
  workflow_call:
    outputs:
      changed_packages:
        description: 'Stringified JSON array of changed packages in this changeset'
        value: ${{ jobs.dry-run.outputs.changed_packages }}
      # AS A CONVENIENCE: These two outputs are shorthand so consumer can just check boolean output for these specific packages
      api_changed:
        description: 'Indicates if the api or its dependencies have changed'
        value: ${{ jobs.dry-run.outputs.api_changed }}
      ui_changed:
        description: 'Indicates if the UI or its dependencies have changed'
        value: ${{ jobs.dry-run.outputs.ui_changed }}

jobs:
  dry-run:
    runs-on: ubuntu-latest
    env:
      # The turbo filter here varies depending on if we're using this workflow in a PR or on a push to a branch
      # For PRs, we want to use `github.event.pull_request.base.sha` to tell turbo to see which packages changed since that SHA
      # For a branch push/merges, the above sha isn't available, so instead, we reference `github.event.before` to determine the previous `HEAD` of the branch we just pushed onto
      #   See SO answer for more info https://stackoverflow.com/a/61861763/2379922
      TURBO_REF_FILTER: ${{ github.event_name == 'pull_request' && github.event.pull_request.base.sha || github.event.before }}
    outputs:
      # Defining a job output for used by the next jobs:
      # https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idoutputs
      changed_packages: ${{ steps.output-changed-packages.outputs.changed_packages }}
      api_changed: ${{ steps.api-changed.outputs.result }}
      ui_changed: ${{ steps.ui-changed.outputs.result }}

    steps:
      - uses: actions/checkout@v3
        with:
          # see `fetch-depth` in README: https://github.com/actions/checkout#checkout-v3, 
          # we set to `0` so the referenced commits are available for the command below
          fetch-depth: 0

      - name: Changeset
        id: changeset
        # 1. We need the 'output' of a turbo dry-run to get a json with all affected packages of these run.
        # 2. The multi line json string is wrapped in EOF delimeters to make the GHA happy: https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#multiline-strings
        # 3. We specify the .github/ folder as a dependency here. Just to be safe, if workflows changed, we use that to mean everything has changes pending... to force redeploys

        run: |
          echo 'result<<CHANGESET_DELIMITER' >> $GITHUB_OUTPUT
          echo "$(npx -y turbo build --dry-run=json --filter=...[$TURBO_REF_FILTER]  --global-deps=.github/*)" >> $GITHUB_OUTPUT
          echo 'CHANGESET_DELIMITER' >> $GITHUB_OUTPUT

      - name: Output changed packages
        id: output-changed-packages
        run: |
          echo 'changed_packages<<CHANGED_PACKAGES_DELIMITER' >> $GITHUB_OUTPUT
          echo "${{ toJSON(fromJSON(steps.changeset.outputs.result).packages) }}" >> $GITHUB_OUTPUT
          echo 'CHANGED_PACKAGES_DELIMITER' >> $GITHUB_OUTPUT

      - name: Did the API or its dependencies change?
        id: api-changed
        # 1. Using the 'output' of the 'changeset' step inside the `if` to check if the specific app has changed
        # 2. When `true`, we store it in the 'output' of the current step.
        # NOTE: `api` is the name of my server package
        if: ${{ contains(fromJSON(steps.changeset.outputs.result).packages, 'api') }} 
        run: echo "result=true" >> $GITHUB_OUTPUT

      - name: Did the UI or its dependencies change?
        id: ui-changed
        # 1. Using the 'output' of the 'changeset' step inside the `if` to check if the specific app has changed
        # 2. When `true`, we store it in the 'output' of the current step.
        # NOTE: `ui` is the name of my client package
        if: ${{ contains(fromJSON(steps.changeset.outputs.result).packages, 'ui') }} 
        run: echo "result=true" >> $GITHUB_OUTPUT

```
</details>

Which can be used in another workflow like so: 
<details>
  <summary><code>.github/workflows/cd.yml</code></summary>

```yml
# ... etc...
jobs:
  changed-packages:
    name: Determine which apps changed
    uses: ./.github/workflows/changed-packages.yml

  deploy-api:
    name: Deploy API using a deployment workflow
    needs: [changed-packages]
    if: needs.changed-packages.outputs.api_changed == 'true' # as far as I can tell, GHA outputs are all strings, see https://github.com/actions/runner/issues/1483
    uses: my/server/deploy/workflow

  deploy-ui:
    name: Deploy UI using a deployment workflow
    needs: [changed-packages]
    if: needs.changed-packages.outputs.ui_changed == 'true'
    uses: my/client/deploy/workflow

# ... etc...
```
</details>

**Reply by bobaaaaa** [reply]
That's a cool enhancement. Thx for sharing 👍 

- Yes, `HEAD^` is used in our project because of a 'squash & merge' strategy
- And yes, `origin/main` is hard-coded, because that's the only relevant branch for us

I will try your version in our setup.

**Reply by jplwood** [reply]
I know we have [`--affected`](https://turbo.build/repo/docs/reference/run#--affected) now, but just wanted to follow up again here and say that if you still need check that some named package has changed (as opposed to just running a task for "affected" packages), the above reusable workflow I had above can be simplified using [`turbo-ignore`](https://www.npmjs.com/package/turbo-ignore) ([docs](https://turbo.build/repo/docs/reference/turbo-ignore)). This new workflow, instead of hardcoding some outputs saying if workspaces _x_, _y_ and _z_ changed, this new one just takes in an NPM workspace name, and says if it changed or not:

<details>
  <summary><code>.github/workflows/turbo-changed.yml</code></summary>

```yml
name: Check if a workspace changed

on:
  workflow_call:
    inputs:
      workspace_name:
        description: 'Name of the workspace, aka the app name in the package.json file'
        required: true
        type: string
      # The base here varies depending on if we're using this workflow in a PR or on a push to a branch, or on the creation of a release/tag
      # e.g., For PRs, we want to use `github.event.pull_request.base.sha` to tell turbo to see which packages changed since that SHA
      # e.g., For a branch push/merges, the above sha isn't available, so instead, we reference `github.event.before` to determine the previous `HEAD` of the branch we just pushed onto
      # e.g., For releases, we want to use the previous tag, see `cd-release.yml`
      #   See SO answer for more info https://stackoverflow.com/a/61861763/2379922
      base_ref:
        description: 'The base git ref to use to diff for changes'
        required: true
        type: string
    outputs:
      changed:
        description: 'Did the inputed workspace (i.e., package) change according to turbo-ignore? Return "true" or "false"'
        value: ${{ jobs.changed.outputs.changed }}
jobs:
  changed:
    runs-on: ubuntu-latest
    outputs:
      changed: ${{ steps.changed.outputs.changed }}
    steps:
      - uses: actions/checkout@v4
        with:
          # https://github.com/actions/checkout#checkout-v3, see `fetch-depth`,
          # we set to `0` so the referenced commits are available for the command below
          fetch-depth: 0

      - name: Did workspace change?
        id: changed
        # `turbo-ignore` exits with 0 if workspace didn't change, and a non-zero (error) if it did change
        # evaluate exit code to output true/false if packaged changed.
        run: |
          # disable "exit on failure"
          set +e  
          npx -y turbo-ignore ${{ inputs.workspace_name }} --fallback ${{ inputs.base_ref }}
          if [ $? -eq 0 ]
          then
            echo "changed=false" >> $GITHUB_OUTPUT
          else
            echo "changed=true" >> $GITHUB_OUTPUT
          fi

```
</details>

**The TL;DR being**, it uses `turbo-changed` instead of --dry-run and JSON parsing, and use the exit code to derive if a pkg "changed":
```yml
# disable "exit on failure"
set +e  
npx -y turbo-ignore ${{ inputs.workspace_name }} --fallback ${{ inputs.base_ref }}
if [ $? -eq 0 ]
then
  echo "changed=false" >> $GITHUB_OUTPUT
else
  echo "changed=true" >> $GITHUB_OUTPUT
fi
```
which is used like so:

```yml
# ... ci, on triggered on PR
jobs:
  # Only run ci on mypkg if it, or one of its local dependencies, has changed
  mypkg-changed:
    uses: ./.github/workflows/turbo-changed.yml
    with:
      workspace_name: mypkg # The name of the turbo workspace (i.e., name in package.json)
      base_ref: ${{ github.event.pull_request.base.sha }}

  run-ci-on-mypkg
    needs: [mypkg-changed]
    if: ${{ needs.mypkg-changed.outputs.changed == 'true' }}
    runs-on: ubuntu-latest
    steps:
      #... run ci
```

**Comment by alexbchr** [comment]
We have this use case here in quite a lot of repositories and I ended up writing a [Github Action](https://github.com/marketplace/actions/turbo-changed) just for that. Hopefully turborepo will provide an official action at some point 🤞

**Comment by robingronlund** [comment]
Following up in this discussion since I'm very interested in this. Looks like you all provided solid solutions. However I was wondering since most of the solutions forces me to be explicit about what package I want to check, is there a way to make this dynamic? Like check all my package/apps and see if any has changed. 

Has anything been updated with turborepo that might support this out of the box?

Also sorry for tailing your original discussion @bobaaaaa

**Reply by bobaaaaa** [reply]
No worries. I'm also interested in better solutions

**Reply by weyert** [reply]
I am generating a custom pipeline in Gitlab CI to do something similar based on ideas from Discord conversations in the past

**Reply by dew-dr0p** [reply]
Any update on this yet, dynamically achieving what was done above?

Great work, @bobaaaaa .

**Reply by 0xMurage** [reply]
There is now an **experimental** command to list all affected apps/packages

```shell
turbo ls --affected --output=json
```
https://turbo.build/blog/turbo-2-1-0#run-tasks-for-changed-packages-with---affected

https://turbo.build/repo/docs/reference/ls#--affected

**Reply by douglasg14b** [reply]
affected seems to just list _every_ package, regardless of changes between `main` and `HEAD`

**Reply by 0xMurage** [reply]
@douglasg14b, back then the `ls --affected` command was still experimental, and it "just worked." Nowadays, there are two caveats to the `--affected` flag:

1. Both branches you are comparing have to be checked out.
2. If the checkout is too shallow, then all packages will be considered changed.

All of this is now in the docs https://turborepo.com/docs/reference/run#--affected

**Reply by alexbchr** [reply]
Also, with the `ls` command, it doesn't seem possible to add additional filters or add set a scope. Like list me only affected packages, but which depends on a `build` task might be different from those for a `lint` task. Potentially for build you'll want `affected` packages to include dependent packages, but not for a `lint` task, where you'll potentially just want to get the edited packages (not their dependents if they were not edited).

**Reply by 0xMurage** [reply]
@alexbchr, what you're looking for is `turbo run`, which executes Turbo tasks. It supports everything you need including filters + affected flags : https://turborepo.com/docs/reference/run

**Reply by alexbchr** [reply]
Thanks @0xMurage, this is what I use in our Github Action (see [other reply](https://github.com/vercel/turborepo/discussions/800#discussioncomment-4238541) on this very same discussion) 😅

**Comment by boompttn** [comment]
Skipping CI task: https://turbo.build/repo/docs/core-concepts/monorepos/skipping-tasks

tldr;
`npx turbo-ignore web --task=build`

**Reply by murbanowicz** [reply]
it does not help at all with github actions tbh - there is no way to stop steps from going on with it :/

**Reply by jplwood** [reply]
@murbanowicz I just updated my previous post to show how you can use `turbo-ignore` in a reusable github actions workflow to run/skip subsequent steps/jobs: https://github.com/vercel/turborepo/discussions/800#discussioncomment-11896317
